_posts/2019-08-23-til-ecto-select-merge.md

```diff
@@ -0,0 +1,237 @@
+---
+author: Kate Travers
+author_link: https://github.com/ktravers
+categories: post
+tags: ['ecto']
+date: 2019-08-23
+layout: post
+title: TIL How to Select Merge with Ecto.Query
+excerpt: >
+  Before you reach for adding another association to your schema, consider using `Ecto.Query#select_merge/3` with a virtual field instead.
+---
+
+Working with Elixir and Ecto, I've run into scenarios where I needed to retrieve data from a table plus maybe a field or two from an unassociated table. In the past, whenever this happened, I'd usually spin up something I wasn't totally satisfied with -- maybe updating the schema(s), breaking it up into multiple queries, or building a multi-select statement if I was feeling fancy.
+
+Happily, today I learned there's a better way. You can accomplish that same end result in a single query expression with `Ecto.Query#select_merge/3`.
+
+Let's run through an example to see it in action.
+
+## Setup
+
+Say you work at a school with an admissions department, and you've been tasked with displaying an event log showing all the events related to a given admission, organized into three columns: 1) date, 2) action taken, and 3) who the action was taken by.
+
+| Date     | Action           | Taken By         |
+|----------|------------------|------------------|
+| 8/1/2019 | Student Admitted | Albus Dumbledore |
+
+
+To start with, we have an `AdmissionEvent` schema that looks like this:
+
+```elixir
+defmodule Registrar.Tracking.AdmissionEvent do
+  use Ecto.Schema
+
+  schema "admission_events" do
+    field(:action, :string)
+    field(:admission_id, :integer)
+    field(:admitter_uuid, Ecto.UUID)
+    field(:occurred_at, :naive_datetime)
+  end
+end
+```
+
+...and a User schema that looks like this:
+
+```elixir
+defmodule Registrar.User do
+  use Ecto.Schema
+
+  schema "users" do
+    field(:uuid, Ecto.UUID)
+    field(:full_name, :string)
+  end
+end
+```
+
+The problem here is the admitter's full name lives on `User`, which isn't currently associated with `AdmissionEvent`. So if we did a straight-forward select query, we'd end up with the admission events we need to populate the table, but not the admitters' full names.
+
+```elixir
+defmodule Registrar.Tracking.AdmissionEvent do
+  use Ecto.Schema
+  import Ecto.Query, only: [from: 2]
+  alias Registrar.Tracking.AdmissionEvent
+
+  schema "admission_events" do
+    field(:action, :string)
+    field(:admission_id, :integer)
+    field(:admitter_uuid, Ecto.UUID)
+    field(:occurred_at, :naive_datetime)
+  end
+
+  def for_admission(query \\ AdmissionEvent, admission) do
+    from(ae in query,
+      where: ae.admission_id == ^admission.id,
+      order_by: [desc: ae.occurred_at]
+    )
+  end
+end
+```
+
+```elixir
+# Taking our query function for a spin...
+
+iex> admission = Repo.get(Admission, 1)
```

Comment:
Didn't notice this before, but there isn't a definition for `Admission`
I think the tutorial would be simple if had one 

What do you think?